Vs2013 win64

sofakid opened the pull request:
Build on windows with the MS build chain.

timblechmann commented on lang/LangSource/PyrSlot32.h:
```diff
@@ -35,6 +35,8 @@ A PyrSlot is an 8-byte value which is either a double precision float or a
 #include <cassert>
 #include <cstddef>
 
+typedef long ptr_sized_int;
```
std::intptr_t

timblechmann commented on lang/LangSource/PyrSlot64.h:
```diff
@@ -284,4 +284,10 @@ inline void slotCopy(PyrSlot *dst, const PyrSlot *src, int num)
 }
 
+#ifdef _WIN64
+typedef long long ptr_sized_int;
```
std::intptr_t

timblechmann commented on lang/LangSource/PyrParseNode.h:
```diff
@@ -555,6 +555,6 @@ extern PyrBlock *gCompilingBlock;
 */
 
-#define YYSTYPE long
+#define YYSTYPE ptr_sized_int
```
std::intptr_t

sofakid replied:
even better

timblechmann commented on lang/LangSource/PyrParseNode.cpp:
```diff
@@ -4102,9 +4103,9 @@ void PyrBlockNode::compile(PyrSlot* slotResult)
-			//totalStrings++;
+			static int totalLength = 0, totalStrings = 0;
+			totalLength += stringLength;
+			totalStrings++;
```
debugging cruft?

sofakid replied:
yeah

timblechmann commented on lang/LangSource/PyrLexer.cpp:
```diff
@@ -1712,11 +1715,12 @@ void compileDepTree()
 		compileClass(classdep->fileSym, classdep->startPos, classdep->endPos, classdep->lineOffset);
 	}
-	//postfl("<compile\n");
+	postfl("<compile\n");
```
debugging cruft

timblechmann commented on lang/LangSource/Bison/lang11d_tab.cpp:
```diff
@@ -4513,7 +4513,7 @@ yyparse ()
 /* Line 1455 of yacc.c  */
 #line 1481 "lang11d"
-    { (yyval) = zzval; ;}
+    { (yyval) = zzval ;}
```
???

sofakid replied:
that's a whoopsiedaisies.

timblechmann commented on include/common/wintime.h:
```diff
@@ -0,0 +1,37 @@
+#include <windows.h>
+
+// HACK - prevent winsock from complaining about duplicate definitions
+#define _WINSOCK2API_
```
doesn't this imply that winsock2.h cannot be included?

sofakid replied:
yeah, from that one file. it is a hack.

timblechmann commented on lang/CMakeLists.txt:
```diff
@@ -293,6 +293,9 @@ endif()
     target_link_libraries(libsclang portmidi)
+    target_link_libraries(libsclang kernel32 user32 gdi32 winspool shell32 
+    	ole32 oleaut32 uuid comdlg32 advapi32 setupapi hidapi hidapi_parser 
+    	libscsynth tlsf boost_thread boost_system winmm yaml)
```
aren't yaml, the boost libs, libscsynth, tlsf, hidapi etc also required for non-windows builds? looks fishy to be, most likely a duplicate. please link libraries only in one location

sofakid replied:
i don't have other systems to test on. those libraries were not being linked on windows.

i was minimizing impact of my changes to the other platforms.

crucialfelix commented on cmake_modules/FindSndfile.cmake:
```diff
@@ -25,14 +25,19 @@ elseif (APPLE)
 else()
 	find_path(SNDFILE_INCLUDE_DIR sndfile.h
-		PATHS "$ENV{ProgramFiles}/Mega-Nerd/libsndfile/include"
+		PATHS "C:/Program Files/Mega-Nerd/libsndfile/include"
```
is this a hard coded path now ? it should stay as an $ENV I think

timblechmann replied:
???

sofakid replied:
I was flailing around, trying to get cmake to find my stuff. 

Cmake is picking the wrong environment variable.

PROGRAMFILES=C:\Program Files\
PROGRAMFILES(X86)=C:\Program Files (x86)\

i guess the brackets confuse cmake and it thinks PROGRAMFILES=C:\Program Files (x86)\

which then links the 32bit library and badness ensues.

sofakid replied:
this else doesn't even run

timblechmann commented:
please do *not* include binaries into the public git repositories! libsndfile.lib and portaudio_x64.lib. they bloat (read: practically slow down) git repositories forever. use zip archives or separate repositories for that.

timblechmann commented:
likewise: do not include upstream headers (sndfile/portaudio) into the project. instead: point the build system to the right places where they can be found. fwiw, libsndfile even have windows installers these days.

timblechmann commented on editors/sc-ide/core/settings/theme.cpp:
```diff
@@ -214,8 +214,14 @@ void Theme::setFormat(const QString & key, const QTextCharFormat & newFormat)
+#ifndef WIN32
+			__func__ <<
+#else
+			"Theme::setFormat" <<
```
__FUNCTION__

sofakid replied:
what's FUNCTION do? (not my code, just didn't work on windows so i took it out on windows.)

timblechmann commented on editors/sc-ide/core/settings/theme.cpp:
```diff
@@ -214,8 +214,14 @@ void Theme::setFormat(const QString & key, const QTextCharFormat & newFormat)
-        qDebug() << __func__ << "Failed to find key " << key;
-        return;
+        qDebug() << 
+#ifndef WIN32
```
WIN32 is not a standard, use _WIN32 instead
http://sourceforge.net/p/predef/wiki/OperatingSystems/

timblechmann commented on QtCollider/primitives/prim_misc.cpp:
```diff
@@ -168,8 +168,11 @@ QC_LANG_PRIMITIVE( Qt_SetGlobalPalette, 1, PyrSlot *r, PyrSlot *a, VMGlobals *g
   if( !QcApplication::compareThread() ) return QtCollider::wrongThreadError();
 
+// VS2013 can't do this.
+#ifndef WIN32
```
_WIN32

timblechmann commented on QtCollider/primitives/prim_misc.cpp:
```diff
@@ -168,8 +168,11 @@ QC_LANG_PRIMITIVE( Qt_SetGlobalPalette, 1, PyrSlot *r, PyrSlot *a, VMGlobals *g
 {
   if( !QcApplication::compareThread() ) return QtCollider::wrongThreadError();
 
+// VS2013 can't do this.
```
why?

sofakid replied:
no idea.

timblechmann commented on .gitignore:
```diff
@@ -20,3 +20,5 @@ common/Packager/SuperCollider*.dmg
 *~
 *.orig
+
+vs
```
?
a proper cmake build system does not generate any file in the source tree

sofakid replied:
that was a folder i had in a different branch and i couldn't get rid of it

bagong commented:
Hey Lucas, thank you so much for this, you're bringing SCWin to a new level!

timblechmann commented:
regarding the mass-renaming from long to ptr_sized_int:

* please use std::intptr_t: http://en.cppreference.com/w/cpp/types/integer
* was this a find/replace? or did you actually analyse the reason for those changes? e.g. it is changing generated headers (bison parser), which looks a bit fishy.

in general, the sizing fix does not look too bad to me, but since it is changing this in *many* locations, i'm a bit worried that it is simply replaced without carefully checking the semantics.

sofakid commented:
There was a bit of find & replacing happening, but i did it after careful examination of the semantics. 

I had to change it in the union too, that wasn't an accident.

sofakid commented:
OK. 

I reverted most of my cmake changes. I'll let Rainer work his magic there. 

I replaced my ptr_sized_int with intptr_t, and regenerated the bison parser. (this isn't a build step.. weird.)

I removed the binaries and headers.

timblechmann commented on CMakeLists.txt:
```diff
@@ -242,6 +242,7 @@ endif()
 
 if(${CMAKE_CXX_COMPILER_ID} STREQUAL "MSVC")
   foreach(flag CMAKE_C_FLAGS CMAKE_CXX_FLAGS)
+		set(${flag} "${${flag}} /wd4065") # switch statement contains 'default' but no 'case' labels
```
please try not to mix spaces and tabs

timblechmann commented on QtCollider/primitives/prim_misc.cpp:
```diff
@@ -168,7 +168,7 @@ QC_LANG_PRIMITIVE( Qt_SetGlobalPalette, 1, PyrSlot *r, PyrSlot *a, VMGlobals *g
   if( !QcApplication::compareThread() ) return QtCollider::wrongThreadError();
 
-  QPalette p = QtCollider::get( a );
+  QPalette p = (QPalette&&) QtCollider::get(a);
```
can you explain this change?

sofakid replied:
I'm not certain about this code. VS was complaining that there were two constructors it could choose to create p, one took a QPalette&&, the other took some kind of button.. It seemed the safest choice.

timblechmann replied:
well, un-safe type-casts are never safe (by definition).

fwiw, this moves (=invalidates) the original object, while the old code copied it.

timblechmann commented on lang/LangSource/PyrLexer.h:
```diff
@@ -22,6 +22,7 @@
 #ifndef _PYRLEXER_H_
 #define _PYRLEXER_H_
 
+#include "PyrSlot.h"
```
is this include necessary?

sofakid replied:
no, it's not, i thought i got all those, sorry.

timblechmann commented on lang/LangSource/PyrSymbol.h:
```diff
@@ -27,6 +27,7 @@ A PyrSymbol is a unique string that resides in a global hash table.
 #define _PYRSYMBOL_H_
 
 #include "SC_Types.h"
+#include "PyrSlot.h"
```
again, is this necessary? afaict, <cstdint> should be sufficient for intptr_t

timblechmann commented on server/scsynth/CMakeLists.txt:
```diff
@@ -187,7 +187,10 @@ else()
-	target_link_libraries(libscsynth wsock32 ws2_32 winmm)
+	target_link_libraries(libscsynth wsock32 ws2_32 winmm kernel32 user32
+        gdi32 winspool shell32 ole32 oleaut32 uuid comdlg32 advapi32 tlsf
+        boost_thread boost_system
```
aren't tlsf, boost_thread and boost_system linked automatically?

timblechmann commented:
thanks a lot for the changes. i only have two remaining points:
* the cast to the rvalue reference looks fishy to me, especially as it is not type-safe, so i wonder, what's the reason for this.
* please don't add and remove the portaudio/sndfile libraries, but drop the commit: git will store them forever in the history, bloating the repo forever.

sofakid commented:
I used the BFG tool to remove the binaries. looks like it deleted it from the past as well. 

https://rtyley.github.io/bfg-repo-cleaner/

timblechmann commented on cmake_modules/FindPortaudio.cmake:
```diff
@@ -39,55 +39,81 @@ else (PORTAUDIO_LIBRARIES AND PORTAUDIO_INCLUDE_DIRS)
-    )
-   
+
+    if (NOT WIN32)
```
is this really necessary?

bagong replied:
Separating the winpaths, you mean? Maybe not, but this way there is more duplication, the other way there would be more confusing if-nesting. Is it a big issue? I can change it if you think it's important.

timblechmann replied:
you can simply add the windows paths or library names. the code is already looking for portaudio_x64, which is a windows library name.

fwiw, duplicating code is a bad thing, as it leads to bitrot and higher maintenance overhead in the long term.

bagong replied:
Done, thanks!
I hope I didn't introduce an error ;)

timblechmann commented on cmake_modules/FindPortaudio.cmake:
```diff
@@ -39,55 +39,81 @@ else (PORTAUDIO_LIBRARIES AND PORTAUDIO_INCLUDE_DIRS)
+            PATHS ${CMAKE_SOURCE_DIR}/../portaudio/include
+        )
+        find_library(PORTAUDIO_LIBRARY
+            NAMES portaudio_x64.lib portaudio_x86.lib libportaudio.a
```
why look for windows libraries in the non-windows codepath?

bagong replied:
Not exactly sure what you mean. AFAIK there is no binary distribution for portaudio and no default install location. So in the readme I suggest a simple tree structure in which sc sourcetree and external dependencies reside as siblings in a parent folder. If you are referring to libportaudio.a, that's for the msys/mingw build. The background is that I am trying to keep the cmake arguments to a minimum.

timblechmann replied:
ah, missed the NOT.

if you want to keep cmake arguments to a minimum, add portaudio to the external libraries and build while building sc.

bagong replied:
Yea, I'll consider that when I have more time. Or how about a pull request? :) I think you did more or less that in the past once and then removed it again... :) 
A nice side effect would be that the guaranteed correct build type will be built each time. Atm if somebody forgets to reconfigure the build as required (clean & delete cmake cache) she might end up with a build type mismatch between portaudio and sc.

bagong commented:
This PR is outdated. I'll close it for now and we'll present something else soon. The current version is here: branch `winport` on https://github.com/bagong/supercollider.git
